dvc/repo/plot/data.py

```diff
@@ -0,0 +1,261 @@
+        self.filename = filename
+        self.revision = revision
+        self.content = content
+
+    def raw(self, **kwargs):
+        raise NotImplementedError
+
+    def _processors(self):
+        return [_filter_fields]
+
+    def to_datapoints(self, **kwargs):
+        data = self.raw(**kwargs)
+
+        for data_proc in self._processors():
+            data = data_proc(
+                data, filename=self.filename, revision=self.revision, **kwargs
+            )
+
+        for data_point in data:
+            data_point[self.REVISION_FIELD] = self.revision
+        return data
+
+
+class JSONPlotData(PlotData):
+    def raw(self, **kwargs):
+        return json.loads(self.content, object_pairs_hook=OrderedDict)
+
+    def _processors(self):
+        parent_processors = super(JSONPlotData, self)._processors()
+        return [_apply_path, _find_data] + parent_processors
+
+
+class CSVPlotData(PlotData):
+    def __init__(self, filename, revision, content, delimiter=","):
+        super(CSVPlotData, self).__init__(filename, revision, content)
+        self.delimiter = delimiter
+
+    def raw(self, csv_header=True, **kwargs):
+        first_row = first(csv.reader(io.StringIO(self.content)))
+
+        if csv_header:
+            reader = csv.DictReader(
+                io.StringIO(self.content), delimiter=self.delimiter,
+            )
+        else:
+            reader = csv.DictReader(
+                io.StringIO(self.content),
+                delimiter=self.delimiter,
+                fieldnames=[str(i) for i in range(len(first_row))],
+            )
+
+        fieldnames = reader.fieldnames
+        data = [row for row in reader]
+
+        return [
+            OrderedDict([(field, data_point[field]) for field in fieldnames])
+            for data_point in data
+        ]
+
+
+class YAMLPLotData(PlotData):
+    def raw(self, **kwargs):
+        return yaml.parse(io.StringIO(self.content))
+
+
+def _load_from_revision(repo, datafile, revision):
+    if revision is WORKSPACE_REVISION_NAME:
+
+        def open_datafile():
+            return repo.tree.open(datafile, "r")
+
+    else:
+
+        def open_datafile():
+            from dvc import api
+
+            return api.open(datafile, repo.root_dir, revision)
+
+    try:
+        with open_datafile() as fobj:
+            datafile_content = fobj.read()
+
+    except (FileNotFoundError, PathMissingError):
+        raise NoMetricOnRevisionError(datafile, revision)
+
+    return plot_data(datafile, revision, datafile_content)
+
+
+def _load_from_revisions(repo, datafile, revisions):
+    data = []
+    exceptions = []
+
+    for rev in revisions:
+        try:
+            data.append(_load_from_revision(repo, datafile, rev))
+        except NoMetricOnRevisionError as e:
+            exceptions.append(e)
+        except PlotMetricTypeError:
+            raise
+        except Exception:
+            logger.error("Failed to parse '{}' at '{}'.".format(datafile, rev))
```

Comment:
Not sure about this one:
```suggestion
            logger.error("Failed to parse '{}' in '{}'.".format(datafile, rev))
```

Reply:
Hmm, isn't it `at revision`?

Reply:
I had no idea what words were around "at" haha. But also, no, it's probably "in". Obvi not important in the logger message but maybe in https://github.com/iterative/dvc/pull/3577#discussion_r416936489